### PR TITLE
Prune vesting accounts balances

### PR DESCRIPTION
**Pull request by alpe**
Prevent an out of gas error when burning accounts balance

Update:
This PR also includes 
* pruning vesting accounts original denoms only
* drop the prune account types list. It is easier to check for the vesting account interface. This should cover future vesting types as well. Chains can still overwrite this behaviour with a custom CoinPruner via Options

### CODE REVIEW

**Comment by alpe** [comment]
An alternative solution would be to burn only the `stakingKeeper.BondDenom()` balance when pruning an account.

File: x/wasm/keeper/keeper.go [hunk 1]
```diff
@@ -1161,30 +1144,38 @@ func (c BankCoinTransferrer) TransferCoins(parentCtx sdk.Context, fromAddr sdk.A
...
-		}
+func (b VestingCoinBurner) PruneBalances(ctx sdk.Context, existingAcc authtypes.AccountI) error {
+	v, ok := existingAcc.(vestingexported.VestingAccount)
+	if !ok {
+		return types.ErrAccountExists.Wrapf("refusing to overwrite special account type:: %T", existingAcc)
+	}
+
+	ctx = ctx.WithGasMeter(sdk.NewInfiniteGasMeter())
+	coinsToBurn := sdk.NewCoins()
+	for _, orig := range v.GetOriginalVesting() { // focus on the coin denoms that were setup originally; getAllBalances has some issues
```
Comment (ethanfrey) [inline x/wasm/keeper/keeper.go:1171]: Nice addition. This is limited to the number of coins that we set in the original vesting creation transaction, and thus must be some reasonable number.

**Review by ethanfrey — COMMENTED** [review]
Logic seems correct.

Hard to follow, I would like PruneBalances to only take vesting account and make that check explicit in instantiate

File: x/wasm/keeper/keeper.go [hunk 2]
```diff
@@ -321,17 +306,15 @@ func (k Keeper) instantiate(ctx sdk.Context, codeID uint64, creator, admin sdk.A
 		if _, accept := k.acceptedAccountTypes[reflect.TypeOf(existingAcct)]; accept {
 			// keep account and balance as it is
 			k.Logger(ctx).Info("instantiate contract with existing account", "address", contractAddress.String())
-		} else if _, clear := k.pruneAccountTypes[reflect.TypeOf(existingAcct)]; clear {
-			k.Logger(ctx).Info("pruning existing account for contract instantiation", "address", contractAddress.String())
+		} else {
```
[the remark this answers is not in the document]
Comment (alpe) [reply x/wasm/keeper/keeper.go:309]: This looks like a bigger change to the previous version but reduces complexity and maintainability a lot by pushing the "prune list" handling one level down into the (custom) CoinPruner implementation. This was designed with extensibility in mind. It is easier to have custom logic in code than to maintain a list of types. I also noticed that the prune list does not add value when I have to cast for the vesting account interface anyway. I hope this helps with my motivation.

File: x/wasm/keeper/keeper.go [hunk 3]
```diff
@@ -1161,30 +1144,38 @@ func (c BankCoinTransferrer) TransferCoins(parentCtx sdk.Context, fromAddr sdk.A
...
-	if amt := b.bank.GetAllBalances(ctx, address); !amt.IsZero() {
-		if err := b.bank.SendCoinsFromAccountToModule(ctx, address, types.ModuleName, amt); err != nil {
-			return sdkerrors.Wrap(err, "prune account balance")
-		}
-		if err := b.bank.BurnCoins(ctx, types.ModuleName, amt); err != nil {
-			return sdkerrors.Wrap(err, "burn account balance")
-		}
+func (b VestingCoinBurner) PruneBalances(ctx sdk.Context, existingAcc authtypes.AccountI) error {
+	v, ok := existingAcc.(vestingexported.VestingAccount)
+	if !ok {
```
[the remark this answers is not in the document]
Comment (alpe) [reply x/wasm/keeper/keeper.go:1165]: This was built with the assumption that also other than SDK account types can exists in a chain. The `CoinPruner.PruneBalances` method is an extension point that can be used so that chains can handle their account types and/or vesting account types as they want. Another implementation could also be sending the tokens to the community pool or an escrow (not saying this makes sense).  
The concrete struct type is `VestingCoinBurner`. I thought this verbose enough and explicit addressing the SDK vesting account types only (or return error).
I can add some more code doc to this but switching to VestingAccount types only would not work due to the custom account types. WDYT?

File: x/wasm/keeper/keeper.go [hunk 4]
```diff
@@ -1161,30 +1144,38 @@ func (c BankCoinTransferrer) TransferCoins(parentCtx sdk.Context, fromAddr sdk.A
...
+		return types.ErrAccountExists.Wrapf("refusing to overwrite special account type:: %T", existingAcc)
+	}
+
+	ctx = ctx.WithGasMeter(sdk.NewInfiniteGasMeter())
+	coinsToBurn := sdk.NewCoins()
+	for _, orig := range v.GetOriginalVesting() { // focus on the coin denoms that were setup originally; getAllBalances has some issues
+		coinsToBurn = coinsToBurn.Add(b.bank.GetBalance(ctx, existingAcc.GetAddress(), orig.Denom))
```
[the remark this answers is not in the document]
Comment (alpe) [reply x/wasm/keeper/keeper.go:1172]: 👍 good eye

File: x/wasm/keeper/keeper.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ethanfrey) [reply x/wasm/keeper/keeper.go:1170]: Ah, okay, so this PruneBalances is not on the Keeper, but can easily be set by an external app. Now I understand the design.

> Throwing types.ErrAccountExists though would better fit one level where it was before. I will modify the method signature to return a handled bool for this

No need for `handled` unless you prefer it as well, I think returning `types.ErrAccountExists` is nice

[the remark this answers is not in the document]
Comment (ethanfrey) [reply x/wasm/keeper/keeper.go:1170]: You convinced me that we need to keep the logic here. That does make sense.

However, the name is misleading. `PruneBalances` is what you do when you encounter a vesting account.

Maybe `CleanupExistingAccount` would be a better name (as is more general and fits what you want as an extension point). And if it is a vesting account, you `PruneBalances`. Maybe other accounts have different cleanup methods? And if there is no cleanup possible (existing module account), returning some `types.ErrCannotReuseAccount` error or such sounds good.

**Review by ethanfrey — COMMENTED** [review]
Okay, I understand your motivation better.
Left comments for alternate cleanup. Basically just name changes now.

File: x/wasm/keeper/keeper.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (alpe) [reply x/wasm/keeper/keeper.go:1170]: 👍 `CleanupExistingAccount` makes sense as a more general term. I prefer the `handled` result value, too as it brings back the control flow to the caller.

**Comment by alpe** [comment]
I have pushed the renaming. Please approve the PR when they make sense now

File: x/wasm/keeper/keeper_test.go [hunk 5]
```diff
@@ -2274,3 +2274,66 @@ func TestAppendToContractHistory(t *testing.T) {
 	gotHistory := keepers.WasmKeeper.GetContractHistory(ctx, contractAddr)
 	assert.Equal(t, orderedEntries, gotHistory)
 }
+
+func TestCoinBurnerPruneBalances(t *testing.T) {
```
Comment (pinosu) [inline x/wasm/keeper/keeper_test.go:2278]: nice test!

**Review by pinosu — APPROVED** [review]
LGTM 👍

**Review by ethanfrey — APPROVED** [review]
looks good now
